File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
    Q := []*TreeNode{root}

    for len(Q)>0 {
        c := Q[0]
        Q = Q[1:]
        if c!=nil {
            Q = append(Q, c.Left)
            Q = append(Q, c.Right)
        }
        if checkEqual(c, subRoot) {
            return true
        }
    }
    return false
}

func checkEqual(root, subRoot *TreeNode) bool {
    if (root == nil && subRoot == nil) {
        return true
    }

    return root != nil && subRoot != nil && 
        root.Val == subRoot.Val &&
        checkEqual(root.Left, subRoot.Left) && 
        checkEqual(root.Right, subRoot.Right)
}

/*
Q = {1,2,3}
c = 1
sr = 2

*/

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        def isBalanced(node: Optional[TreeNode], height: int) -> (int, bool):
            if not node:
                return (height, True)
            
            leftHeight, leftBalanced = isBalanced(node.left, height)
            rightHeight, rightBalanced = isBalanced(node.right, height)

            balanced = leftBalanced and rightBalanced and abs(leftHeight - rightHeight)<=1
            return (max(leftHeight, rightHeight)+1, balanced)

        return isBalanced(root, 0)[1]
        